cmd/containerd-shim-runc-v2/task/service.go

```diff
@@ -707,18 +730,94 @@ func (s *service) send(evt interface{}) {
-		if runc.ShouldKillAllOnExit(s.context, c.Bundle) {
-			if err := ip.KillAll(s.context); err != nil {
-				log.G(s.context).WithError(err).WithField("id", ip.ID()).
-					Error("failed to kill init's children")
+// handleInitExit processes container init process exits.
+// This is handled separately from non-container init exits, because there
+// are some extra invariants we want to ensure in this case, namely:
+// - the init process exit MUST be the last exit we publish for a given container
+// - IF a container init process is reaped, we MUST ALWAYS publish it's exit
+// This is achieved by:
+// - waiting for all pending execs to either successfully start or early-exit
+// (this happens before handleInitExit is called)
+// - killing all running container processes
+// - waiting for the running process count (for a container) to go to 0, or
+// timing out.
+// - removing all processes from s.running so that we don't publish late
+// exec exits, and publishing the init exit.
+func (s *service) handleInitExit(e runcC.Exit, c *runc.Container, p *process.Init) {
+	s.lifecycleMu.Lock()
+	defer s.lifecycleMu.Unlock()
+
+	// kill all running container processes
+	if err := p.KillAll(s.context); err != nil {
+		log.G(s.context).WithError(err).WithField("id", p.ID()).
+			Error("failed to kill init's children")
+	}
+
+	// subscribe to exits so that we can receive exits for our running processes
+	exits := make(map[int][]runcC.Exit)
+	s.exitSubscribers[&exits] = struct{}{}
+
+	// remove all running processes from s.running, so that their exits don't get
+	// processed by processExits
+	runningCps := make(map[int][]containerProcess)
+	for _, execProcess := range c.ExecdProcesses() {
+		pid := execProcess.Pid()
+		if _, running := s.running[execProcess.Pid()]; running {
+			var skipped []containerProcess
+			for _, cp := range s.running[pid] {
+				if cp.Container == c {
+					runningCps[pid] = append(runningCps[pid], cp)
+				} else {
+					skipped = append(skipped, cp)
+				}
+			}
+
+			if len(skipped) == 0 {
+				delete(s.running, pid)
+			} else {
+				s.running[pid] = skipped
 			}
 		}
 	}
 
+	go func() {
+		execExits := make(map[int][]runcC.Exit)
+		for { // wait for running processes to exit
+			<-time.After(20 * time.Millisecond)
+
+			var updated bool
+			for pid := range runningCps {
+				if e, exited := exits[pid]; exited {
+					execExits[pid] = append(execExits[pid], e...)
+					updated = true
+				}
+			}
+
+			if !updated {
+				break
+			}
+
+			if len(execExits) == len(runningCps) {
+				break
+			}
+		}
+
+		for pid, execExits := range execExits {
+			for _, exit := range execExits {
+				for _, cp := range runningCps[pid] {
+					s.handleProcessExit(exit, cp.Container, cp.Process)
+				}
+			}
+		}
+
+		// all running processes have exited now, and no new
+		// ones can start, so we can publish the init exit
+		s.handleProcessExit(e, c, p)
+	}()
+}
+
+// s.mu must be locked when calling handleProcessExit
```

Comment:
[Lies!](https://github.com/containerd/containerd/pull/9104#issuecomment-1737902483)

Reply:
Yup! I noticed that (and should've removed it). It's not even been the case that we actually hold `s.mu` while calling that for a while 😅